bot/exts/info/doc/_cog.py

```diff
@@ -0,0 +1,427 @@
+        """
+        Rebuild the inventory for a single package.
+
+        Where:
+            * `package_name` is the package name to use, appears in the log
+            * `base_url` is the root documentation URL for the specified package, used to build
+                absolute paths that link to specific symbols
+            * `inventory_url` is the absolute URL to the intersphinx inventory.
+        """
+        self.base_urls[api_package_name] = base_url
+
+        for group, items in package.items():
+            for symbol, relative_doc_url in items:
+
+                # e.g. get 'class' from 'py:class'
+                group_name = group.split(":")[1]
+                if (original_symbol := self.doc_symbols.get(symbol)) is not None:
+                    replaced_symbol_name = self.ensure_unique_symbol_name(
+                        api_package_name,
+                        group_name,
+                        original_symbol,
+                        symbol,
+                    )
+                    if replaced_symbol_name is not None:
+                        symbol = replaced_symbol_name
+
+                relative_url_path, _, symbol_id = relative_doc_url.partition("#")
+                # Intern fields that have shared content so we're not storing unique strings for every object
+                symbol_item = DocItem(
+                    api_package_name,
+                    sys.intern(group_name),
+                    base_url,
+                    sys.intern(relative_url_path),
+                    symbol_id
+                )
+                self.doc_symbols[symbol] = symbol_item
+                self.item_fetcher.add_item(symbol_item)
+
+        log.trace(f"Fetched inventory for {api_package_name}.")
+
+    async def update_or_reschedule_inventory(
+            self,
+            api_package_name: str,
+            base_url: str,
+            inventory_url: str
+    ) -> None:
+        """
+        Update the cog's inventory, or reschedule this method to execute again if the remote inventory unreachable.
+
+        The first attempt is rescheduled to execute in `FETCH_RESCHEDULE_DELAY.first` minutes, the subsequent attempts
+        in `FETCH_RESCHEDULE_DELAY.repeated` minutes.
+        """
+        package = await fetch_inventory(inventory_url)
+
+        if not package:
+            attempt = self.inventory_reschedule_attempts[package]
+            self.inventory_reschedule_attempts[package] += 1
+            if attempt == 0:
+                delay = FETCH_RESCHEDULE_DELAY.first
+            else:
+                delay = FETCH_RESCHEDULE_DELAY.repeated
+            log.info(f"Failed to fetch inventory; attempting again in {delay} minutes.")
+            self.inventory_scheduler.schedule_later(
+                delay*60,
+                (attempt, api_package_name),
+                self.update_or_reschedule_inventory(api_package_name, base_url, inventory_url)
+            )
```

Comment:
Since the scheduler shields the scheduled coroutine, I believe the following code will work. It's simpler than keeping track of attempts yourself. 

Since `refresh_inventory` cancels all tasks at the start, I don't think it'd be possible for this to cancel anything but itself. There may be an edge case where the task scheduled by an older invocation of `refresh_inventory` takes a (very) long time to cancel and the new `refresh_inventory` invocation manages to get far enough to schedule a new task for the same package. That would be solved by making `cancel_all` await all tasks it cancels so it blocks until they're cancelled. I believe I discussed that in a previous comment for something unrelated.

```suggestion
            if api_package_name in self.inventory_scheduler:
                self.inventory_scheduler.cancel(api_package_name)
                delay = FETCH_RESCHEDULE_DELAY.repeated
            else:
                delay = FETCH_RESCHEDULE_DELAY.first

            log.info(f"Failed to fetch inventory; attempting again in {delay} minutes.")
            self.inventory_scheduler.schedule_later(
                delay*60,
                api_package_name,
                self.update_or_reschedule_inventory(api_package_name, base_url, inventory_url)
            )
```